pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,244 @@
+
+References are the only exception to top down evaluation. By definition,
+references (both static and dynamic) make the evaluation process break from
+this mode by <i>jumping</i> to another location (as we saw in the evaluation path
+figure at the beginning of the Dynamic Scope section). As a consequence, when
+the evaluation process encounters a reference, lexical scopes and dynamic
+scopes will typically diverge.
+
+To recap, for the lexical scope, following a reference involves abandoning the
+lexical scope of the reference schema and entering the lexical scope of the
+destination schema. In comparison, for the dynamic scope, <i>following a
+reference involves retaining the current dynamic scope, and adding the
+destination schema to it</i>.
+
+Consider the following example of a schema that follows a reference to an
+anchor. In the left, the lexical scope of the destination schema (highlighted
+in red), consists of it parent schema (highlighted in green), and its
+grandparent schema (highlighted in blue). However, in the right, the dynamic
+scope of the destination schema (highlighted in red) consists of the schema
+that referenced the anchor (highlighted in green) and the schema that led to
+the evaluation of such reference (highlighted in blue):
+
+![Comparison of lexical scope and dynamic scope with a reference](/img/posts/2024/understanding-lexical-dynamic-scopes/references-dynamic.webp)
+
+### Beyond Schema Resource Boundaries
+
+In comparison to the lexical scope, the dynamic scope of a schema is not
+limited by its schema resource. Moreover, the dynamic scope is a singleton
+entity that remains for the entirety of the evaluation process. In other words,
+the dynamic scope <i>is</i> the complete evaluation path.
+
+Consider a variation of the previous example, where this time the anchor is
+within a subschema of a different schema resource. The lexical scope of the
+destination schema drastically changes as it hits its schema resource barrier,
+where as the dynamic scope of the destination schema remains the same:
+
+![Comparison of lexical scope and dynamic scope through schema resources](/img/posts/2024/understanding-lexical-dynamic-scopes/references-dynamic-resource.webp)
+
+Finally, note that the dynamic scope not only persists through local references
+to other schema resources, but across remote references as well.
+
+Summary
+-------
+
+Understanding how static and dynamic scopes work is essential for gaining a
+deeper understanding of JSON Schema. The most important points to remember are
+summarized in the following table:
+
+| Comparison Point                     | Lexical Scope | Dynamic Scope |
+|--------------------------------------|---------------|---------------|
+| **Definition**                       | Consists of the parent schemas including the schema itself, but not its subschemas | Consists of the subschemas that were evaluated so far, including the schema itself, but not its subschemas |
```

Comment:
Again, not sure how this definition of lexical scope is beneficial beyond distinguishing it from dynamic scope.

Reply:
The point of the article was to exemplify lexical & dynamic scopes by themselves, not necessarily showcasing any specific use cases (i.e. we don't dive into dynamic referencing, which is the only use case of dynamic scopes at this point if I understand correctly).

Is there another kind of definition that you have in place? Might be related to your Draft 7 comment above, which I don't think I fully understand yet. Can you provide more info on this? What kind of definition would you expect?